infra/src/Pos/Infra/InjectFail.hs

```diff
@@ -0,0 +1,84 @@
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE OverloadedStrings          #-}
+-- | Fault injection, strategically placed at the root of CSL dependency graph.
+
+module Pos.Infra.InjectFail
+       ( FInjects
+       , parseFInjectsSpec
+       , mkFInjects
+       , FInject(..)
+       , testFInject
+       , setFInject
+       , logFInject
+       , testLogFInject
+       ) where
+
+import           Universum
+
+import           Data.Aeson (FromJSON, ToJSON)
+import           Data.Char (toLower)
+import           Data.Maybe (catMaybes)
+import           Data.Set (Set)
+import qualified Data.Set as Set
+import qualified Data.Text as T
+import           Options.Applicative
+
+import           Pos.Util.Wlog (CanLog, HasLoggerName, logError,
+                     modifyLoggerName)
+
+
+
+data FInject
+  = FInjIgnoreAPI                     -- ^ Return a hard-coded string for all registered Wallet API endpoints
+  | FInjIgnoreShutdown                -- ^ Ignore the shutdown request
+  | FInjApplyUpdateNoExit             -- ^ Don't exit after handling the 'update/apply' endpoint
+  | FInjApplyUpdateWrongExitCode      -- ^ Exit with a wrong exit code as response to the 'update/apply' request
+  deriving (Bounded, Enum, Eq, Generic, Ord, Show)
+instance FromJSON FInject
+instance ToJSON   FInject
+
+desc :: FInject -> String
+desc = \case
+    FInjIgnoreAPI                     ->  "Return a hard-coded string for all registered Wallet API endpoints"
+    FInjIgnoreShutdown                ->  "Ignore the shutdown request"
+    FInjApplyUpdateNoExit             ->  "Don't exit after handling the 'update/apply' endpoint"
+    FInjApplyUpdateWrongExitCode      ->  "Exit with a wrong exit code as response to the 'update/apply' request"
+
+newtype FInjects = FInjects { _fromFInjs :: IORef (Set FInject) }
+
+parseFInjectsSpec :: Parser (Set FInject)
+parseFInjectsSpec = Set.fromList . catMaybes <$> sequenceA (parseSingle <$> enumFromTo minBound maxBound)
+
+parseSingle :: FInject -> Parser (Maybe FInject)
+parseSingle fi = (<&> \case
+                     Just True -> Just fi
+                     _         -> Nothing) <$> optional $
+  switch (long (drop 4 $ map toLower $ show fi) <>
+          help (desc fi))
```

Comment:
That's a hard one to read :sweat_smile: ... Took me a while to also understand the type signature and that the FInject was just given as a way to extract the description and parameter's name! May I simply suggest something along those lines:

```hs
parseSingle :: FInject -> Parser (Maybe FInject)
parseSingle fi =
    flag Nothing (Just fi) opt
  where
    opt = long (drop 4 $ map toLower $ show fi) <> help (desc fi)
```

Reply:
Done!